File: backend/app/routes/pipeline.py
"""
ApeAI — Pipeline Routes

Endpoints for triggering AI pipeline stages:
- /pipeline/cluster
- /pipeline/summarize/{cluster_id}
- /pipeline/generate-brd/{cluster_id}
- /pipeline/generate-prd/{cluster_id}
- /pipeline/generate-stories/{cluster_id}
- /pipeline/generate-tasks/{story_id}
"""

import logging
from typing import Dict, Any, List, Optional

from fastapi import APIRouter, HTTPException

from backend.app.ai.services.clustering_service import cluster_unprocessed_feedback
from backend.app.ai.services.generation_service import (
    summarize_cluster,
    generate_brd,
    generate_prd,
    generate_stories,
    generate_tasks
)

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/pipeline", tags=["AI Pipeline Orchestration"])

@router.post("/cluster", summary="Cluster all unprocessed feedback")
async def api_cluster_feedback():
    """Triggers the iterative clustering logic for all un-clustered feedback."""
    try:
        return await cluster_unprocessed_feedback()
    except Exception as e:
        logger.error(f"Clustering failed: {e}")
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/summarize/{cluster_id}", summary="Generate AI summary for a cluster")
async def api_summarize_cluster(cluster_id: str):
    """Analyzes feedback in a cluster and updates its title and summary."""
    try:
        return await summarize_cluster(cluster_id)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except Exception as e:
        logger.error(f"Summarization failed: {e}")
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/generate-brd/{cluster_id}", summary="Generate BRD for a cluster")
async def api_generate_brd(cluster_id: str):
    """Generates a Business Requirements Document based on the cluster summary."""
    try:
        return await generate_brd(cluster_id)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        logger.error(f"BRD generation failed: {e}")
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/generate-prd/{cluster_id}", summary="Generate PRD for a cluster")
async def api_generate_prd(cluster_id: str, brd_id: str):
    """Generates a Product Requirements Document based on a BRD."""
    try:
        return await generate_prd(cluster_id, brd_id)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        logger.error(f"PRD generation failed: {e}")
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/generate-stories/{cluster_id}", summary="Generate User Stories for a cluster")
async def api_generate_stories(cluster_id: str, prd_id: str):
    """Generates Agile User Stories based on a PRD."""
    try:
        return await generate_stories(cluster_id, prd_id)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        logger.error(f"Stories generation failed: {e}")
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/generate-tasks/{story_id}", summary="Generate Tasks for a story")
async def api_generate_tasks(cluster_id: str, story_id: str):
    """Breaks down a User Story into technical tasks."""
    try:
        return await generate_tasks(cluster_id, story_id)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        logger.error(f"Task generation failed: {e}")
        raise HTTPException(status_code=500, detail=str(e))

@router.get("/status", summary="Full pipeline status")
async def api_pipeline_status():
    """Returns an overview of where all feedback is in the pipeline."""
    # This calls the existing document service helper
    from backend.app.services.document_service import get_pipeline_status
    return await get_pipeline_status()

File: backend-node/src/routes/pipeline.ts
/**
 * ApeAI — Pipeline Routes (Layer 3 — AI Orchestration)
 *
 * GET  /pipeline/status                       — Full pipeline overview
 * POST /pipeline/cluster                      — Cluster all unprocessed feedback
 * POST /pipeline/summarize/:clusterId         — Generate AI summary
 * POST /pipeline/generate-brd/:clusterId      — Generate BRD
 * POST /pipeline/generate-prd/:clusterId      — Generate PRD
 * POST /pipeline/generate-stories/:clusterId  — Generate User Stories
 * POST /pipeline/generate-tasks/:storyId      — Generate Tasks
 *
 * Mirrors Python routes/pipeline.py exactly.
 */

import { Router, Request, Response, NextFunction } from 'express';
import { requireAuth } from '../middleware/auth';
import { getPipelineStatus } from '../services/documentService';
import { clusterUnprocessedFeedback } from '../ai/clusteringService';
import {
  summarizeCluster,
  generateBrd,
  generatePrd,
  generateStories,
  generateTasks,
} from '../ai/generationService';

export const pipelineRouter = Router();

// GET /pipeline/status
pipelineRouter.get(
  '/status',
  async (_req: Request, res: Response, next: NextFunction) => {
    try {
      const status = await getPipelineStatus();
      res.json(status);
    } catch (err) {
      next(err);
    }
  },
);

// POST /pipeline/cluster  (auth required)
pipelineRouter.post(
  '/cluster',
  requireAuth,
  async (req: Request, res: Response, next: NextFunction) => {
    try {
      const result = await clusterUnprocessedFeedback(req.userId);
      res.json(result);
    } catch (err) {
      console.error('Clustering failed:', err);
      next(err);
    }
  },
);

// POST /pipeline/summarize/:clusterId
pipelineRouter.post(
  '/summarize/:clusterId',
  async (req: Request, res: Response, next: NextFunction) => {
    try {
      const result = await summarizeCluster(req.params['clusterId']!);
      res.json(result);
    } catch (err) {
      if (err instanceof Error && err.message.includes('not found')) {
        res.status(404).json({ detail: err.message });
        return;
      }
      console.error('Summarization failed:', err);
      next(err);
    }
  },
);

// POST /pipeline/generate-brd/:clusterId
pipelineRouter.post(
  '/generate-brd/:clusterId',
  async (req: Request, res: Response, next: NextFunction) => {
    try {
      const result = await generateBrd(req.params['clusterId']!);
      res.json(result);
    } catch (err) {
      if (err instanceof Error && (err.message.includes('must be summarized') || err.message.includes('not found'))) {
        res.status(400).json({ detail: err.message });
        return;
      }
      console.error('BRD generation failed:', err);
      next(err);
    }
  },
);

// POST /pipeline/generate-prd/:clusterId?brd_id=
pipelineRouter.post(
  '/generate-prd/:clusterId',
  async (req: Request, res: Response, next: NextFunction) => {
    const brdId = req.query['brd_id'] as string;
    if (!brdId) {
      res.status(400).json({ detail: "'brd_id' query parameter is required" });
      return;
    }
    try {
      const result = await generatePrd(req.params['clusterId']!, brdId);
      res.json(result);
    } catch (err) {
      if (err instanceof Error && err.message.includes('not found')) {
        res.status(400).json({ detail: err.message });
        return;
      }
      console.error('PRD generation failed:', err);
      next(err);
    }
  },
);

// POST /pipeline/generate-stories/:clusterId?prd_id=
pipelineRouter.post(
  '/generate-stories/:clusterId',
  async (req: Request, res: Response, next: NextFunction) => {
    const prdId = req.query['prd_id'] as string;
    if (!prdId) {
      res.status(400).json({ detail: "'prd_id' query parameter is required" });
      return;
    }
    try {
      const result = await generateStories(req.params['clusterId']!, prdId);
      res.json(result);
    } catch (err) {
      if (err instanceof Error && err.message.includes('not found')) {
        res.status(400).json({ detail: err.message });
        return;
      }
      console.error('Stories generation failed:', err);
      next(err);
    }
  },
);

// POST /pipeline/generate-tasks/:storyId?cluster_id=
pipelineRouter.post(
  '/generate-tasks/:storyId',
  async (req: Request, res: Response, next: NextFunction) => {
    const clusterId = req.query['cluster_id'] as string;
    if (!clusterId) {
      res.status(400).json({ detail: "'cluster_id' query parameter is required" });
      return;
    }
    try {
      const result = await generateTasks(clusterId, req.params['storyId']!);
      res.json(result);
    } catch (err) {
      if (err instanceof Error && err.message.includes('not found')) {
        res.status(400).json({ detail: err.message });
        return;
      }
      console.error('Task generation failed:', err);
      next(err);
    }
  },
);
